File: packages/ui/src/data/donut.math.ts
/** Donut arithmetic, kept pure so the awkward inputs are asserted (ADR 0012). */

export interface DonutSlice {
  value: number;
  label: string;
}

export interface DonutArc extends DonutSlice {
  /** Fraction of the whole, 0–1. */
  fraction: number;
  /** Stroke dash length for this arc, in circumference units. */
  length: number;
  /** Stroke dash offset that positions it after the arcs before it. */
  offset: number;
  index: number;
}

/**
 * Turns values into arcs. Negative values are dropped rather than drawn backwards, and a set
 * that sums to zero produces no arcs at all instead of dividing by it.
 */
export function donutArcs(slices: DonutSlice[], circumference: number): DonutArc[] {
  const usable = slices.filter((slice) => Number.isFinite(slice.value) && slice.value > 0);
  const total = usable.reduce((sum, slice) => sum + slice.value, 0);
  if (total <= 0) return [];

  let consumed = 0;
  return usable.map((slice, index) => {
    const fraction = slice.value / total;
    const length = fraction * circumference;
    const arc: DonutArc = { ...slice, fraction, length, offset: -consumed, index };
    consumed += length;
    return arc;
  });
}

File: packages/ui_flutter/lib/src/data/mini_charts/donut_math.dart
import 'package:flutter/foundation.dart';

/// One named part of a [CbDonut].
@immutable
class CbDonutSlice {
  const CbDonutSlice({required this.value, required this.label});

  final double value;
  final String label;

  @override
  bool operator ==(Object other) =>
      other is CbDonutSlice && other.value == value && other.label == label;

  @override
  int get hashCode => Object.hash(value, label);
}

/// The proportional geometry of one drawable [CbDonutSlice].
@immutable
class CbDonutArc {
  const CbDonutArc({
    required this.value,
    required this.label,
    required this.fraction,
    required this.length,
    required this.offset,
    required this.index,
  });

  final double value;
  final String label;
  final double fraction;
  final double length;
  final double offset;
  final int index;

  @override
  bool operator ==(Object other) =>
      other is CbDonutArc &&
      other.value == value &&
      other.label == label &&
      other.fraction == fraction &&
      other.length == length &&
      other.offset == offset &&
      other.index == index;

  @override
  int get hashCode =>
      Object.hash(value, label, fraction, length, offset, index);
}

/// Turns positive finite values into adjacent proportional arcs.
///
/// Negative and non-finite values are dropped rather than painted backwards. A
/// zero total returns no arcs, which avoids division by zero.
List<CbDonutArc> cbDonutArcs(
  Iterable<CbDonutSlice> slices,
  double circumference,
) {
  final List<CbDonutSlice> usable = slices
      .where((CbDonutSlice slice) => slice.value.isFinite && slice.value > 0)
      .toList(growable: false);
  final double total = usable.fold(
    0,
    (double sum, CbDonutSlice slice) => sum + slice.value,
  );
  if (total <= 0) return const <CbDonutArc>[];

  double consumed = 0;
  return <CbDonutArc>[
    for (final (int index, CbDonutSlice slice) in usable.indexed)
      (() {
        final double fraction = slice.value / total;
        final double length = fraction * circumference;
        final CbDonutArc arc = CbDonutArc(
          value: slice.value,
          label: slice.label,
          fraction: fraction,
          length: length,
          offset: -consumed,
          index: index,
        );
        consumed += length;
        return arc;
      })(),
  ];
}
